message_passing/waiting_for_handlers_and_compensation/workflows.py

```diff
@@ -0,0 +1,180 @@
+                # to exit.
+                return await update_task
+            else:
+                # 👉 The main workflow method exited prematurely due to an
+                # error, and this happened before the update finished. Fail the
+                # update with the workflow exception as cause.
+                raise exceptions.ApplicationError(
+                    "The update failed because the workflow run exited"
+                ) from cast(BaseException, self.workflow_exit.exception())
+        # 👉 Catch BaseException since asyncio.CancelledError does not inherit
+        # from Exception.
+        except BaseException as e:
+            if is_workflow_exit_exception(e):
+                try:
+                    await self.my_update_compensation()
+                except BaseException as e:
+                    raise exceptions.ApplicationError(
+                        "Update compensation failed"
+                    ) from e
+            raise
+
+    async def my_update_compensation(self):
+        await workflow.execute_activity(
+            activity_executed_by_update_handler_to_perform_compensation,
+            start_to_close_timeout=timedelta(seconds=10),
+        )
+        self._update_compensation_done = True
+
+    @workflow.query
+    def workflow_compensation_done(self) -> bool:
+        return self._workflow_compensation_done
+
+    @workflow.query
+    def update_compensation_done(self) -> bool:
+        return self._update_compensation_done
+
+    # The following methods are placeholders for the actual application logic
+    # that you would perform in your main workflow method  or update handler.
+    # Their implementation can be ignored.
+
+    async def _my_update(self) -> str:
+        # Ignore this method unless you are interested in the implementation
+        # details of this sample.
+        self._update_started = True
+        await workflow.execute_activity(
+            activity_executed_by_update_handler,
+            start_to_close_timeout=timedelta(seconds=10),
+        )
+        return "update-result"
+
+    async def _run(self, input: WorkflowInput) -> str:
+        # Ignore this method unless you are interested in the implementation
+        # details of this sample.
+
+        # Wait until handlers have started, so that we are demonstrating that we
+        # wait for them to finish.
+        await workflow.wait_condition(lambda: self._update_started)
+        if input.exit_type == WorkflowExitType.SUCCESS:
+            return "workflow-result"
+        elif input.exit_type == WorkflowExitType.FAILURE:
+            raise exceptions.ApplicationError("deliberately failing workflow")
+        elif input.exit_type == WorkflowExitType.CANCELLATION:
+            # Block forever; the starter will send a workflow cancellation request.
+            await asyncio.Future()
+        raise AssertionError("unreachable")
+
+
+def is_workflow_exit_exception(e: BaseException) -> bool:
```

Comment:
👍 , This feels like a concept worth highlighting (and maybe adding a helper?) that I think lots of users would benefit from.  Good to take every opportunity to teach users about which exceptions fail the workflow and the fact that some don't fail it (and that this is a good thing), and then show how that intersects with cleanup code like this.